src/gui/guiEditBox.cpp

```diff
@@ -93,3 +101,733 @@ void GUIEditBox::setAutoScroll(bool enable)
+				if (m_mark_begin == m_mark_end)
+					new_mark_begin = m_cursor_pos;
+				new_mark_end = p;
+			} else {
+				new_mark_begin = 0;
+				new_mark_end = 0;
+			}
+			m_cursor_pos = p;
+			m_blink_start_time = porting::getTimeMs();
+		} break;
+		case KEY_RETURN:
+			if (m_multiline) {
+				inputChar(L'\n');
+			} else {
+				calculateScrollPos();
+				sendGuiEvent(EGET_EDITBOX_ENTER);
+			}
+			return true;
+		case KEY_LEFT:
+
+			if (event.KeyInput.Shift) {
+				if (m_cursor_pos > 0) {
+					if (m_mark_begin == m_mark_end)
+						new_mark_begin = m_cursor_pos;
+
+					new_mark_end = m_cursor_pos - 1;
+				}
+			} else {
+				new_mark_begin = 0;
+				new_mark_end = 0;
+			}
+
+			if (m_cursor_pos > 0)
+				m_cursor_pos--;
+			m_blink_start_time = porting::getTimeMs();
+			break;
+		case KEY_RIGHT:
+			if (event.KeyInput.Shift) {
+				if (Text.size() > (u32)m_cursor_pos) {
+					if (m_mark_begin == m_mark_end)
+						new_mark_begin = m_cursor_pos;
+
+					new_mark_end = m_cursor_pos + 1;
+				}
+			} else {
+				new_mark_begin = 0;
+				new_mark_end = 0;
+			}
+
+			if (Text.size() > (u32)m_cursor_pos)
+				m_cursor_pos++;
+			m_blink_start_time = porting::getTimeMs();
+			break;
+		case KEY_UP:
+			if (m_multiline || (m_word_wrap && m_broken_text.size() > 1)) {
+				s32 lineNo = getLineFromPos(m_cursor_pos);
+				s32 mb = (m_mark_begin == m_mark_end)
+							 ? m_cursor_pos
+							 : (m_mark_begin > m_mark_end ? m_mark_begin
+										      : m_mark_end);
+				if (lineNo > 0) {
+					s32 cp = m_cursor_pos -
+						 m_broken_text_positions[lineNo];
+					if ((s32)m_broken_text[lineNo - 1].size() < cp)
+						m_cursor_pos = m_broken_text_positions
+									       [lineNo - 1] +
+							       core::max_((u32)1,
+									       m_broken_text[lineNo -
+											       1]
+											       .size()) -
+							       1;
+					else
+						m_cursor_pos = m_broken_text_positions
+									       [lineNo - 1] +
+							       cp;
+				}
+
+				if (event.KeyInput.Shift) {
+					new_mark_begin = mb;
+					new_mark_end = m_cursor_pos;
+				} else {
+					new_mark_begin = 0;
+					new_mark_end = 0;
+				}
+			} else {
+				return false;
+			}
+			break;
+		case KEY_DOWN:
+			if (m_multiline || (m_word_wrap && m_broken_text.size() > 1)) {
+				s32 lineNo = getLineFromPos(m_cursor_pos);
+				s32 mb = (m_mark_begin == m_mark_end)
+							 ? m_cursor_pos
+							 : (m_mark_begin < m_mark_end ? m_mark_begin
+										      : m_mark_end);
+				if (lineNo < (s32)m_broken_text.size() - 1) {
+					s32 cp = m_cursor_pos -
+						 m_broken_text_positions[lineNo];
+					if ((s32)m_broken_text[lineNo + 1].size() < cp)
+						m_cursor_pos = m_broken_text_positions
+									       [lineNo + 1] +
+							       core::max_((u32)1,
+									       m_broken_text[lineNo +
+											       1]
+											       .size()) -
+							       1;
+					else
+						m_cursor_pos = m_broken_text_positions
+									       [lineNo + 1] +
+							       cp;
```

Comment:
What's this formatting? I'd prefer to keep the old style with `// clang-format off` "guards" than this. Same above.

Reply:
didn't saw that crap omg :)

Reply:
maybe too many indent i may use a function instead i think